File: examples/coagents-research-canvas/agent-python/research_canvas/delete.py
"""Delete Resources"""

import json
from typing import cast
from research_canvas.state import AgentState
from langchain_core.runnables import RunnableConfig
from langchain_core.messages import ToolMessage, AIMessage
async def delete_node(state: AgentState, config: RunnableConfig): # pylint: disable=unused-argument
    """
    Delete Node
    """
    return state

async def perform_delete_node(state: AgentState, config: RunnableConfig): # pylint: disable=unused-argument
    """
    Perform Delete Node
    """
    ai_message = cast(AIMessage, state["messages"][-2])
    tool_message = cast(ToolMessage, state["messages"][-1])
    if tool_message.content == "YES":
        if ai_message.tool_calls:
            urls = ai_message.tool_calls[0]["args"]["urls"]
        else:
            parsed_tool_call = json.loads(ai_message.additional_kwargs["function_call"]["arguments"])
            urls = parsed_tool_call["urls"]

        state["resources"] = [
            resource for resource in state["resources"] if resource["url"] not in urls
        ]

    return state

File: examples/canvas/research-canvas/agents/typescript/src/delete.ts
/**
 * Delete Resources
 */

import type { AgentState } from "./state";
import type { RunnableConfig } from "@langchain/core/runnables";
import type { ToolMessage, AIMessage } from "@langchain/core/messages";

export async function delete_node(
  state: AgentState,
  config: RunnableConfig,
): Promise<AgentState> {
  /**
   * Delete Node
   */
  return state;
}

export async function perform_delete_node(
  state: AgentState,
  config: RunnableConfig,
) {
  /**
   * Perform Delete Node
   */
  const aiMessage = state["messages"][
    state["messages"].length - 2
  ] as AIMessage;
  const toolMessage = state["messages"][
    state["messages"].length - 1
  ] as ToolMessage;

  let resources = state["resources"];

  if (toolMessage.content === "YES") {
    let urls: string[];

    if (aiMessage.tool_calls) {
      urls = aiMessage.tool_calls[0].args.urls;
    } else {
      const parsedToolCall = JSON.parse(
        aiMessage.additional_kwargs!.function_call!.arguments,
      );
      urls = parsedToolCall.urls;
    }

    resources = resources.filter((resource) => !urls.includes(resource.url));
  }

  return {
    resources,
  };
}
